### PR TITLE
Fix K-Cache Fill Pruning for Accesses Outside of Full Loop Interval

### LINKED ISSUE #438: Vertical stencil broken in GTC CUDA backends (gtc:cuda and gtc:gt:gpu)
This stencil seems to confuse the `gtc:cuda` backend:
```python
with interval(2, -1):
    denratm = grg[0, 0, -1] / grg[0, 0, 0]
    ee[0,0,0] = (1.0-ee[0,0,-2])
```

Generated C++

```c++
double ee1m2;
double ee1m1;
double ee1p0;

double grg1m1;
double grg1p0;

// VerticalLoopSection 140572861651312
for (int _k_block = 2; _k_block < 3; ++_k_block) {

// HorizontalExecution 140572861487424
if (validator(extent<0, 0, 0, 0>())) {

  grg1m1 = grg(0_c, 0_c, -1_c);
  grg1p0 = grg(0_c, 0_c, 0_c);
  denratm(0_c, 0_c, 0_c) = (grg1m1 / grg1p0);
  ee1p0 = (static_cast<double>(1.0) - ee1m2);
  ee(0_c, 0_c, 0_c) = ee1p0;
}

  sid::shift(_ptr, sid::get_stride<dim::k>(m_strides), 1_c);

  ee1m2 = ee1m1;
  ee1m1 = ee1p0;

  grg1m1 = grg1p0;
}

// VerticalLoopSection 140572861282096
for (int _k_block = 3; _k_block < k_size + -1; ++_k_block) {

  // HorizontalExecution 140572861283152
  if (validator(extent<0, 0, 0, 0>())) {

    grg1p0 = grg(0_c, 0_c, 0_c);
    denratm(0_c, 0_c, 0_c) = (grg1m1 / grg1p0);
    ee1p0 = (static_cast<double>(1.0) - ee1m2);
    ee(0_c, 0_c, 0_c) = ee1p0;
  }

  sid::shift(_ptr, sid::get_stride<dim::k>(m_strides), 1_c);

  ee1m2 = ee1m1;
  ee1m1 = ee1p0;

  grg1m1 = grg1p0;
}
```

### PR DESCRIPTION
**Pull request by fthaler**
## Description

The PruneKCacheFills optimization pass wrongly assumed that no accesses to fill fields happen outside of the loop interval. Fixes #438.

## Requirements

Before submitting this PR, please make sure:

- [x] The code builds cleanly without new errors or warnings
- [x] The code passes all the existing tests
- [x] If this PR adds a new feature, new tests have been added to test these
new features
- [x] All relevant documentation has been updated or added

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

**Comment by fthaler** [comment]
bors try

File: src/gtc/passes/oir_optimizations/caches.py [hunk 1]
```diff
@@ -186,6 +186,20 @@ def requires_fill(field: str) -> bool:
             return {field for field in filling_fields if not requires_fill(field)}
 
         pruneable = set.intersection(*(pruneable_fields(section) for section in node.sections))
+
+        # Consider accesses outside of loop interval
+        # Note: those accesses would not require a fill in the whole loop interval in in theory
```
Comment (havogt) [inline src/gtc/passes/oir_optimizations/caches.py:191]:
```suggestion
        # Note: those accesses would not require a fill in the whole loop interval in theory
```

Comment (havogt) [reply src/gtc/passes/oir_optimizations/caches.py:191]: I know it's annoying, but please expand the comment with details, maybe with example, such that a non-GridTools developers can understand.

**Review by havogt — CHANGES_REQUESTED** [review]

File: src/gtc/passes/oir_optimizations/caches.py (hunk 1 above)
Comment (fthaler) [reply src/gtc/passes/oir_optimizations/caches.py:191]: Done. Hope that’s understandable enough.

Comment (havogt) [reply src/gtc/passes/oir_optimizations/caches.py:191]: time will tell ;)

**Review by havogt — APPROVED** [review]

**Pull request merged**